src/core/config.c

```diff
@@ -28,63 +30,57 @@ void finish_config(struct xdpw_config *config) {
 		return;
 	}
-	const char *c = iniparser_getstring(d, key, fallback);
-	if (c == NULL) {
-		return;
-	}
-	// Allow keys without value as default
-	if (strcmp(c, "") != 0) {
-		*dest = strdup(c);
-	} else {
-		*dest = fallback ? strdup(fallback) : NULL;
-	}
+	free(*dest);
+	*dest = strdup(value);
 }
 
-static void getdouble_from_conffile(dictionary *d,
-		const char *key, double *dest, double fallback) {
-	if (*dest != 0) {
+static void parse_double(double *dest, const char* value) {
+	if (value == NULL || *value == '\0') {
+		logprint(TRACE, "config: skipping empty value in config file");
 		return;
 	}
-	*dest = iniparser_getdouble(d, key, fallback);
-}
-
-static bool file_exists(const char *path) {
-	return path && access(path, R_OK) != -1;
+	*dest = strtod(value, (char**)NULL);
 }
 
-static void config_parse_file(const char *configfile, struct xdpw_config *config) {
-	dictionary *d = NULL;
-	if (configfile) {
-		logprint(INFO, "config: using config file %s", configfile);
-		d = iniparser_load(configfile);
-	} else {
-		logprint(INFO, "config: no config file found");
-	}
-	if (configfile && !d) {
-		logprint(ERROR, "config: unable to load config file %s", configfile);
-	}
-
-	// screencast
-	getstring_from_conffile(d, "screencast:output_name", &config->screencast_conf.output_name, NULL);
-	getdouble_from_conffile(d, "screencast:max_fps", &config->screencast_conf.max_fps, 0);
-	getstring_from_conffile(d, "screencast:exec_before", &config->screencast_conf.exec_before, NULL);
-	getstring_from_conffile(d, "screencast:exec_after", &config->screencast_conf.exec_after, NULL);
-	getstring_from_conffile(d, "screencast:chooser_cmd", &config->screencast_conf.chooser_cmd, NULL);
-	if (!config->screencast_conf.chooser_type) {
+static int ini_parser_handler(void *data, const char* section, const char* name, const char* value) {
+	struct xdpw_config *config = (struct xdpw_config*)data;
+	logprint(TRACE, "config: parsing setction %s, key %s, value %s", section, name, value);
+
+	#define MATCH(s, n) strcmp(section, s) == 0 && strcmp(name, n) == 0
+	if (MATCH("screencast", "output_name")) {
+		parse_string(&config->screencast_conf.output_name, value);
+	} else if (MATCH("screencast", "max_fps")) {
+		parse_double(&config->screencast_conf.max_fps, value);
+	} else if (MATCH("screencast", "exec_before")) {
+		parse_string(&config->screencast_conf.exec_before, value);
+	} else if (MATCH("screencast", "exec_after")) {
+		parse_string(&config->screencast_conf.exec_after, value);
+	} else if (MATCH("screencast", "chooser_cmd")) {
+		parse_string(&config->screencast_conf.chooser_cmd, value);
+	} else if (MATCH("screencast", "chooser_type")) {
 		char *chooser_type = NULL;
-		getstring_from_conffile(d, "screencast:chooser_type", &chooser_type, "default");
+		parse_string(&chooser_type, value);
 		config->screencast_conf.chooser_type = get_chooser_type(chooser_type);
 		free(chooser_type);
+	} else {
+		logprint(TRACE, "config: skipping invalid key in config file");
```

Comment:
Making this an error would make it easier for users to understand why one of their config options doesn't work (e.g. typo). This is kind of unrelated to this PR though.

Reply:
This only works nice together with [`logprint(TRACE, "config: parsing setction %s, key %s, value %s", section, name, value);`](https://github.com/emersion/xdg-desktop-portal-wlr/pull/126/files/1b7b22ec2665cdd11c02ae96ec2df79cf6cef8f2#diff-4189891b601c97c24b3f16f7fea437d38af30d2d1c34fc8b692e7b96b153a37cR52)